distributed/worker_state_machine.py

```diff
@@ -1292,11 +1269,95 @@ def _ensure_communicating(self, *, stimulus_id: str) -> RecsInstrs:
                 )
             )
 
-        for ts in skipped_worker_in_flight_or_busy:
-            self.data_needed.add(ts)
+            self.comm_nbytes += total_nbytes
+            self.in_flight_workers[worker] = to_gather_keys
+            if (
+                len(self.in_flight_workers) >= self.total_out_connections
+                and self.comm_nbytes >= self.comm_threshold_bytes
+            ):
+                break
 
         return recommendations, instructions
 
+    def _select_workers_for_gather(self) -> Iterator[tuple[str, HeapSet[TaskState]]]:
+        """Helper of _ensure_communicating.
+
+        Yield the peer workers and tasks in data_needed, sorted by:
+
+        1. first local, then remote
+        2. if tied, by highest-priority task available
+        3. if tied, by number of tasks available to be fetched (see note below)
+        4. if tied, by a random element
+
+        Omit workers that are either busy or in flight.
+        Remove peer workers with no tasks from data_needed.
+
+        Note
+        ----
+        Instead of number of tasks, we could've measured total nbytes and/or number of
+        tasks that only exist on the worker. Raw number of tasks is cruder but simpler.
+        """
+        host = get_address_host(self.address)
+        heap = []
+
+        for worker, tasks in list(self.data_needed.items()):
+            if not tasks:
+                del self.data_needed[worker]
+                continue
+            if worker in self.in_flight_workers or worker in self.busy_workers:
+                continue
```

Comment:
I know I initially suggested to go down this road but I would still like to raise the point that this implementation now would iterate over all workers every time we're calling `_ensure_communicating`

Reply:
Even worse, if I'm not mistaken, the below iteration is not even linear (even without the additional push).

Reply:
The iteration is O(n) to the number of peer workers that hold any data to fetch, and O(n*logn) to the number of workers that hold data to fetch and are neither in flight nor busy - split between O(n) in python bytecode * O(logn) in C.

In a worst-case scenario of a cluster with 1000 workers where you suddenly have data to fetch from all 1000 (and I seriously doubt this is realistic), you'd perform 
1. a for loop in pure python of 1000 iterations
2. a single `heapq.heapify` on a list of 1000 elements, which costs O(n*logn) and is implemented in C
3. a for loop in pure python of 50 iterations (`distributed.worker.connections.outgoing`), each of which calls `heapq.heappop` on a list of 950~1000 elements, which costs O(logn) and is implemented in C

At the next call to _ensure_communicating, before any worker has responded, you'll just repeat step 1, while skipping steps 2 and 3.

I strongly suspect this whole thing to be negligible in most cases.

*I can easily improve this, moving aside the busy/in flight workers to a separate attribute (e.g. WorkerState.busy_workers), thus making the second call O(1).* However if you agree I'd rather do it in a follow-up PR.